### PR TITLE
Huobi fetchBorrowRates

### PR DESCRIPTION
**Pull request by samgermain**
So huobi's api response is kind of weird, it returns something like

```
"data": [
    {
      "symbol": "1inchusdt",
      "currencies": [
        {
          "currency": "1inch",
          "interest-rate": "0.00098",
          "min-loan-amt": "90.000000000000000000",
          "max-loan-amt": "1000.000000000000000000",
          "loanable-amt": "0.0",
          "actual-rate": "0.00098"
        },
        {
          "currency": "usdt",
          "interest-rate": "0.00098",
          "min-loan-amt": "100.000000000000000000",
          "max-loan-amt": "1000.000000000000000000",
          "loanable-amt": "0.0",
          "actual-rate": "0.00098"
        }
      ]
    },
    {
      "symbol": "aaveusdt",
      "currencies": [
        {
          "currency": "aave",
          "interest-rate": "0.00098",
          "min-loan-amt": "1.500000000000000000",
          "max-loan-amt": "30.000000000000000000",
          "loanable-amt": "0.0",
          "actual-rate": "0.00098"
        },
        {
          "currency": "usdt",
          "interest-rate": "0.00098",
          "min-loan-amt": "100.000000000000000000",
          "max-loan-amt": "8000.000000000000000000",
          "loanable-amt": "0.0",
          "actual-rate": "0.00098"
        }
      ]
    },
 ```
 
 so I just assign a dictionary value to a response, and if it shows up again it's overwritten
 On my account the interest rate values are the same for every pair, and they're also the same for every coin, always 0.00098
 Before I write `fetchBorrowRate` I watch to know if this method is good
 
 --------------------------
 
# Response

```
   ...
            ZIL: {  currency: "ZIL",
                  rate:  0.00098,
                  span:  86400000,
             timestamp:  1637650709614,
              datetime: "2021-11-23T06:58:29.614Z" },
      ZRX: {  currency: "ZRX",
                  rate:  0.00098,
                  span:  86400000,
             timestamp:  1637650709614,
              datetime: "2021-11-23T06:58:29.614Z" }  }
```

### CODE REVIEW

**Comment by samgermain** [comment]
The other exchanges had the key `info` which I would be worried about here because the values would be different depending on the market

```
          "min-loan-amt": "100.000000000000000000",
          "max-loan-amt": "1000.000000000000000000",
          "loanable-amt": "0.0",
```

**Comment by kroitor** [comment]
I think we should reconsider the name of this method or the format of borrow rates, cause in this case it has to be indexed by symbol.

File: js/huobi.js [hunk 1]
```diff
@@ -2447,6 +2448,31 @@ module.exports = class huobi extends Exchange {
...
+        const rates = {};
+        for (let i = 0; i < data.length; i++) {
+            const market = data[i];
+            const currencies = this.safeValue (market, 'currencies');
+            for (let j = 0; j < currencies.length; j++) {
+                const currency = currencies[j];
+                const currencyId = this.safeString (currency, 'currency');
+                const upperCurrencyId = this.safeCurrencyCode (currencyId, 'currency');
+                rates[upperCurrencyId] = {
+                    'currency': upperCurrencyId,
```
Comment (kroitor) [inline js/huobi.js:2465]:
```JavaScript
                const upperCurrencyId = this.safeCurrencyCode (currencyId, 'currency');
                rates[upperCurrencyId] = {
                    'currency': upperCurrencyId,
```
↓
```JavaScript
                const code = this.safeCurrencyCode (currencyId, 'currency');
                rates[code] = {
                    'currency': code,
```

File: js/huobi.js [hunk 2]
```diff
@@ -2447,6 +2448,31 @@ module.exports = class huobi extends Exchange {
         };
     }
 
+    async fetchBorrowRates (params = {}) {
+        await this.loadMarkets ();
+        const response = await this.privateGetMarginLoanInfo (params);
```
Comment (kroitor) [inline js/huobi.js:2453]: We will need to use the new `spot` API endpoint here, `spotPrivateGetV1MarginLoanInfo`. The old endpoints are deprecated: 

- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L109
- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L252

**Comment by samgermain** [comment]
> I think we should reconsider the name of this method or the format of borrow rates, cause in this case it has to be indexed by symbol, then by currency.

Yeah but this is the only exchange that is like that, the other 4 exchanges give you the interest rate for a borrowed currency. I'm trying to create that functionality on huobi, so that people can accurately calculate the interest that would be charged for a margin trade. 

It looks like every single coin has an interest rate of `0.00098`, so I don't know if this method is useful even, other than to provide a unified structure if someone wanted to use this method across multiple exchanges

**Comment by kroitor** [comment]
> Yeah but this is the only exchange that is like that, the other 4 exchanges give you the interest rate for a borrowed currency. 

Instead of trying to bring it all to a single method, we can have a family of related methods. Basically, the point of unification is to have methods with the same names return the same resulting structures. Also, if two methods have the same name, their unified return structures should not differ.

> I'm trying to create that functionality on huobi, so that people can accurately calculate the interest that would be charged for a margin trade.

Given the following conditions:

(A) One or more exchanges return one borrow rate per currency, for one currency at a time.
(B) One or more exchanges return borrow rates by currency, for all currencies at once.
(C) One or more exchanges return borrow rates by currency by symbol, for all symbols at once.

An exchange can belong to groups A, B, C (either one, or two or all of them).

Underlining the above is the required superset of functionality that we need to cover all possible cases A, B, C.

Is is up to you to devise the proper set of methods to cover the above functionality.

I was thinking of something like:

A = `fetchBorrowRate`
B = `fetchBorrowRates`
C = `fetchBorrowRatesPerSymbol` (?)

In general, `fetchBorrowRate` (by one currency code) can be emulated via `fetchBorrowRates` (by all currency codes) in the base classes in `Exchange.js`, `exchange.py`, `Exchange.php` in the following way:

```JavaScript
async fetchBorrowRate (code, params = {})) {
    await exchange.loadMarkets ();
    if (!this.has['fetchBorrowRates']) {
        throw new NotSupported (this.id + 'fetchBorrowRate() is not supported yet')
    }
    const borrowRates = await this.fetchBorrowRates (params);
    const rate = this.safeValue (borrowRates, code);
    if (rate === undefined) {
        throw new ExchangeError (this.id + 'fetchBorrowRate() could not find currency code ' + code);
    }
    return rate;
}
```

> It looks like every single coin has an interest rate of 0.00098, so I don't know if this method is useful even, other than to provide a unified structure if someone wanted to use this method across multiple exchanges

Since Huobi's exchange-wide margin rate is 0.00098 according to their fee-schedule https://www.huobi.com/en-us/fee/, Huobi can belong to all groups, A, B and C. First, we implement `fetchBorrowRatesPerSymbol` which should return something like:

```JavaScript
{
    'BTC/USDT': {
        'BTC': { ... borrow rate structure ...},
        'USDT': { ... borrow rate structure ...}
    },
    'ETH/USDT': {
        'BTC': { ... borrow rate structure ...},
        'USDT': { ... borrow rate structure ...}
    },
}
```

Then we emulate `fetchBorrowRates` in `huobi.js` based on `fetchBorrowRatesPerSymbol`:

```JavaScript
async fetchBorrowRates (params = {})) {
    await exchange.loadMarkets ();
    const borrowRatesPerSymbol = await this.fetchBorrowRatesPerSymbol (params);
    const symbols = Object.keys (borrowRatesPerSymbol);
    const result = {};
    for (let i = 0; i < symbols.length; i++) {
        const symbol = symbols[i];
        const borrowRates = this.safeValue (borrowRatesPerSymbol, symbol);
        const codes = Object.keys (borrowRates);
        for (let j = 0; j < codes.length; j++) {
            result[code] = borrowRates[code];
        }
    }
    return result;
}
```

The (optional) emulation of `fetchBorrowRate` in the base class is built on top of `fetchBorrowRates`.

Thus we can have a superset of possible methods that covers all exchanges to this or that extent. Lmk if that makes sense.

P.S. You can apply the same unifying-generalizing approach in all similar situations and cases.

**Comment by samgermain** [comment]
I'm getting this result with FTX for some reason 

```
node examples/js/cli.js ftx fetchBorrowRate USDT
Debugger listening on ws://127.0.0.1:54861/02d00d34-06b0-4dd7-83c7-1b631648886e
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
2021-11-30T05:04:36.109Z
Node.js: v14.17.0
CCXT v1.62.56
ftx.fetchBorrowRate (USDT)
---------------------------------------------------
[TypeError] currencyId.toUpperCase is not a function

    at safeCurrency               js/base/Exchange.js:1344            'code': (currencyId === undefined) ? currencyId : this.commonCurrencyCode (curr…
    at safeCurrencyCode           js/base/Exchange.js:1349            currency = this.safeCurrency (currencyId, currency)                             
    at fetchBorrowRate            js/ftx.js:2318                      const coin = this.safeCurrencyCode (rate, 'coin');                              
    at processTicksAndRejections  internal/process/task_queues.js:95                                                                                  
    at async main                 examples/js/cli.js:246              const result = await exchange[methodName] (... args)                            

currencyId.toUpperCase is not a function
```

**Pull request merged**